### ISSUE TITLE
mount option x-systemd.makefs fails due to hardcoded path

### ISSUE THREAD

**Issue by jbgi** [NixOS/nixpkgs#41180]
## Issue description
```
/nix/store/gnz7xgrh04jarccmc2in0panz7c6lxya-systemd-237/lib/systemd/systemd-makefs

systemd-makefs[13822]: /sbin/mkfs.ext4 is not executable: No such file or directory
```
There is probably the same issue with x-systemd.growfs

**Comment by justinkb** [comment]
what you wanna do is add some stuff to the binaryReplacements, to replace the find_executable parameters in src/shared/mkfs-util.c (note you'll need to search for `\"`-escaped, and replace with `\\\"`-escaped strings). you'll end up having to rebuild everything that depends on systemd in some specific way, tho, which is quite a drag (makes it so you basically can't use the nixos cache for loads of packages). I'm gonna submit a PR soon that'll comprehensively tackle this issue, hopefully it can be merged for the next time something else important touches the systemd definition (I'm assuming maintainers don't merge stuff for systemd lightly)

**Comment by flokli** [comment]
I think it might make sense to use the `exec*p` functions instead - see a similar PR at https://github.com/systemd/systemd/pull/27306.

**Comment by justinkb** [comment]
I don't know about that, that change set seems more like something to leave for the systemd folks to figure out, if they wanna apply that logic to the rest of their codebase too. I think minimal changes to `src/shared/mkfs-util.c` should suffice for our purposes.

**Comment by flokli** [comment]
Well, I was the one sending that patch upstream/ opening the PR ;-)

We try to actively remove the number of patches we ship downstream in systemd (https://github.com/NixOS/nixpkgs/issues/80038). Which means, making this a proper patch, patching things the way upstream likes it will reduce the effort to just send it upstream afterwards.

**Comment by justinkb** [comment]
Ah... I did not realize that. I didn't pay attention to the usernames, lol.
I'll take a look at that implementation method, then!

**Comment by justinkb** [comment]
I ran into this for mkswap (from util-linux) when using systemd-cryptsetup
(you should be able to reproduce it with a crypttab entry that specifies a
swap entry), so something like this issue is definitely still relevant. Not
sure why the find_executable invocation for mkswap can't find the one from
util-linux, but it can't.

**Comment by ElvishJerricco** [comment]
This works in stage 1 but not yet in stage 2.

**Comment by ElvishJerricco** [comment]
```nix
import <nixpkgs/nixos/tests/make-test-python.nix> ({ lib, pkgs, ... }: {
  name = "foo";

  nodes.machine = { pkgs, ... }: {
    virtualisation.emptyDiskImages = [ 512 ];
    virtualisation.fileSystems."/foo" = {
      device = "/dev/vdb";
      fsType = "ext4";
      options = ["x-systemd.makefs"];
    };
  };

  testScript = ''
    machine.wait_for_unit("multi-user.target")
    print(machine.succeed("cat /etc/fstab"))
    machine.succeed("mount | grep '/dev/vdb on /foo'")
  '';
})
```

```
machine # [    2.182599] systemd[1]: Starting Make File System on /dev/vdb...
machine # [    2.190177] systemd-makefs[523]: mkfs binary for ext4 is not available.
```

**Comment by flokli** [comment]
Ok, but this is not a hardcoded path in systemd, it's picking up from $PATH:

https://github.com/systemd/systemd/blob/750d9859c139e0fac9d29d3e3147a4367f0fc764/src/shared/mkfs-util.c#L502

Do we maybe not provide the necessary binaries?

**Comment by ElvishJerricco** [comment]
Right, `systemd-makefs@` doesn't have `util-linux` and `fsPackages` on `$PATH`.
